### ISSUE TITLE
IndexedLet should not flag let-vars that correspond to keyword args

### ISSUE THREAD

**Issue by jasonkarns** [rubocop/rubocop-rspec#2002]
## Is your feature request related to a problem? Please describe.

The `IndexedLet` cop is complaining about let-vars that are named specifically to match keyword args (in particular, keyword args in the UUT). Adhering to this cop makes the let-vars inconsistent with the keyword args, and worse, makes it so kwarg shorthand can't be used.

## Describe the solution you'd like

I would like the `IndexedLet` cop to automatically allow (without any configuration) any offending let-vars used as kwarg values. (ie, any let-vars whose name matches a kwarg and would be using kwarg shorthand)

## Describe alternatives you've considered

One option is to rename the kwarg to match the `IndexedLet`-compliant names. However, in many (most?) cases, this would mean that a rubocop-rspec rule is forcing changes to _application code_, which is a tenuous line to cross.

Another option would be to explicitly list the allowed values. However, whether a let-var should be flagged or allowed is context dependent. In one case that maps to kwargs and uses shorthand, they should be allowed; but in another case where they are positional args or any other regular variable reference they should still be flagged. So listing them as exceptions is overly broad.

## Additional context

Example that presently violates `IndexedLet` that should be allowed:

```rb
    let(:line1) { "333 Green St" }
    let(:line2) { "Unit 1000" }
    let(:address) { Address.new(line1:, line2:, city:, state:, zip:) }
```

**Comment by pirj** [comment]
Maybe just allow line1 and line2 in cop config?

**Comment by jasonkarns** [comment]
> Another option would be to explicitly list the allowed values. However, whether a let-var should be flagged or allowed is context dependent. In one case that maps to kwargs and uses shorthand, they should be allowed; but in another case where they are positional args or any other regular variable reference they should still be flagged. So listing them as exceptions is overly broad.

**Comment by pirj** [comment]
It’s impossible to reliably detect if a let-var was ysed anywhere as a kwarg.
Also, what if it’s used in one place as a kwarg, while in another place its usage should be considered as an offence?
What if it’s not used in shorthand syntax?
What if it’s used in shorthand syntax to build a Hash?

Too many factors and cases to consider.

On the other hand, if some project makes extensive use of line1/line2 in specs, it’s easy for them to allowlist those two.

I’m not convinced we should go any further than improving the docs of the cop with an example where line1/line2 are legitimate.
WDYT @rubocop/rubocop-rspec ?

**Comment by bquorning** [comment]
For this to work, we _could_ index all application code and make a list of keyword arguments in all method definitions. But we would probably not capture _all_ method definitions (think `define_method` etc.) and we would still know nothing about methods defined outside the application code (e.g. gem dependencies).

We would need to do a lot more work than we do today, and there would still be false negatives. So no, I don’t think this is something we want to do.

The solution must be to either 1) configure the cop, or 2) disable the cop. And yes, perhaps we should add some more documentation to the cop as @pirj suggests.

**Comment by pirj** [comment]
A PR to improve docs on how to configure the cop is welcome!

**Comment by jasonkarns** [comment]
> For this to work, we could index all application code and make a list of keyword arguments in all method definitions

**Comment by bquorning** [comment]
Hm, yes. Limiting its scope to within the same file might work. Of course there may be edge cases with shared_setup and such, but that may be ok.

I'd be curious to see an implementation, seeing how much extra complexity we'd take on.

And the issue mentioned by @pirj persists: the cop reporting different results for different files may confuse users. (Documentation may fix that.)

**Comment by jasonkarns** [comment]
> And the issue mentioned by @pirj persists: the cop reporting different results for different files may confuse users.

**Comment by pirj** [comment]
You can [configure](https://docs.rubocop.org/rubocop-rspec/latest/cops_rspec.html#configurable-attributes-rspecindexedlet) allowed names and patterns.

**Comment by jasonkarns** [comment]
I don't consider the ability to define exceptions as a valid solution to the core problem.

As mentioned in the original issue description:

> Another option would be to explicitly list the allowed values. However, whether a let-var should be flagged or allowed is context dependent. In one case that maps to kwargs and uses shorthand, they should be allowed; but in another case where they are positional args or any other regular variable reference they should still be flagged. So listing them as exceptions is overly broad.

Ignoring a specific let-var in the _entire test suite_, simply because it exists in one single file as a kwarg is too broad. That can effectively neuter the utility of this cop at all.

**Comment by pirj** [comment]
I see what you mean.
Wouldn't implementing this as you suggest suddenly make the following legitimate?
```ruby
RSpec.describe Registry do
  def enlist(user1:, user2:, user3:)
    [user1, user2, user3].each(&:enlist)
  end

  let(:user1) { ... }
  let(:user2) { ... }
  let(:user3) { ... }

 before { enlist(user1:, user2:, user3:) }
...
```
Pretty far-fetched, just pointing that the same criteria can match a total mess of a use case.

**Comment by jasonkarns** [comment]
Yes it would. (If this hypothetical new mode is used). I think that's the trade off.

If :contextual mode is added in order to reduce false positives, the degree of accuracy is coupled to the degree of complexity in the implementation. Indexing the entire codebase in order to know what values are allowed would reduce false negatives, at the cost of complexity.

My proposal to only inspect kwarg- invocations within the same file is a compromise that reduces false positives, without a significant increase in complexity. At the cost of _potential_ for (likely rare) false negatives.

**Comment by pirj** [comment]
Sounds reasonable.
Would you like to take a stab at it?

> without a significant increase in complexity

I really hope it's achievable.

**Comment by jasonkarns** [comment]
I'd love to but realistically, unlikely. Hopefully someone else can give it a shot?
